File: Difficulty/Medium/[129] Sum Root to Leaf Numbers/solution.cpp
/*
 * @lc app=leetcode id=129 lang=cpp
 *
 * [129] Sum Root to Leaf Numbers
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        if(root == nullptr) return 0;
        queue<TreeNode*> q_tree;
        queue<int> q_val;
        int sum = 0;
        q_tree.push(root);
        q_val.push(root->val);
        while (!q_tree.empty()) {
            int size = q_tree.size();
            while(size--) {
                TreeNode * cur = q_tree.front(); q_tree.pop();
                int cur_val = q_val.front(); q_val.pop();
                if (cur->left != nullptr) {
                    q_tree.push(cur->left);
                    q_val.push(cur->left->val + cur_val * 10);
                }
                if (cur->right != nullptr) {
                    q_tree.push(cur->right);
                    q_val.push(cur->right->val + cur_val * 10);
                }
                if (cur->left == nullptr && cur->right == nullptr)
                    sum += cur_val;
            }
        }
        return sum;
    }
};
// @lc code=end

File: Difficulty/Medium/[129] Sum Root to Leaf Numbers/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from collections import deque

class Solution:
    def sumNumbers(self, root: TreeNode) -> int:
        if root is None: return 0
        q_tree, q_val, res = deque(), deque(), 0
        q_tree.append(root)
        q_val.append(root.val)
        while q_tree:
            size = len(q_tree)
            for _ in range(size):
                cur, cur_val = q_tree.popleft(), q_val.popleft()
                if cur.left:
                    q_tree.append(cur.left)
                    q_val.append(cur.left.val + cur_val * 10)
                if cur.right:
                    q_tree.append(cur.right)
                    q_val.append(cur.right.val + cur_val * 10)
                if not cur.left and not cur.right:
                    res += cur_val
        return res
                
